File: cs/Minsk.Tests/CodeAnalysis/Text/AnnotatedText.cs
using System.Collections.Immutable;
using System.Text;
using Minsk.CodeAnalysis.Text;

namespace Minsk.Tests.CodeAnalysis.Text;

internal sealed class AnnotatedText
{
    private AnnotatedText(string text, ImmutableArray<TextSpan> spans)
    {
        Text = text;
        Spans = spans;
    }

    public string Text { get; }
    public ImmutableArray<TextSpan> Spans { get; }

    public static AnnotatedText Parse(string text)
    {
        text = Unindent(text);

        var textBuilder = new StringBuilder();
        var spanBuilder = ImmutableArray.CreateBuilder<TextSpan>();
        var startStack = new Stack<int>();
        var position = 0;
        foreach (var c in text)
        {
            switch (c)
            {
                case '[':
                    startStack.Push(position);
                    break;
                case ']':
                    if (startStack.Count == 0)
                    {
                        throw new ArgumentException("Too many closing brackets in text", nameof(text));
                    }

                    var start = startStack.Pop();
                    var span = TextSpan.FromBounds(start, position);
                    spanBuilder.Add(span);
                    break;
                default:
                    position++;
                    textBuilder.Append(c);
                    break;
            }
        }

        if (startStack.Count != 0)
        {
            throw new ArgumentException("Too many opening brackets in text", nameof(text));
        }

        return new AnnotatedText(textBuilder.ToString(), spanBuilder.ToImmutable());
    }

    private static string Unindent(string text)
    {
        var lines = UnindentLines(text);
        return string.Join(Environment.NewLine, lines);
    }

    public static List<string> UnindentLines(string text)
    {
        var lines = new List<string>();
        using (var stringReader = new StringReader(text))
        {
            string? line;
            while ((line = stringReader.ReadLine()) != null)
            {
                lines.Add(line);
            }
        }

        if (lines.Count == 0)
        {
            return lines;
        }

        var minIndentation = lines.Where(line => line.Trim().Length != 0)
            .Select(line => line.Length - line.TrimStart().Length).Min();

        for (var i = 0; i < lines.Count; i++)
        {
            if (lines[i].Length < minIndentation)
            {
                lines[i] = "";
                continue;
            }

            lines[i] = lines[i][minIndentation..];
        }

        while (lines.Count > 0 && lines[0].Length == 0)
        {
            lines.RemoveAt(0);
        }

        while (lines.Count > 0 && lines[^1].Length == 0)
        {
            lines.RemoveAt(lines.Count - 1);
        }

        return lines;
    }
}

File: kt/minsk/src/test/kotlin/dev/phytolizer/minsk/analysis/text/AnnotatedText.kt
package dev.phytolizer.minsk.analysis.text

class AnnotatedText private constructor(val text: String, val spans: List<TextSpan>) {
    companion object {
        fun parse(text: String): AnnotatedText {
            val unindentedText = unindent(text)
            val textBuilder = StringBuilder()
            val spanBuilder = mutableListOf<TextSpan>()
            val startStack = mutableListOf<Int>()
            var position = 0
            for (c in unindentedText) {
                when (c) {
                    '[' -> {
                        startStack.add(position)
                    }
                    ']' -> {
                        if (startStack.size == 0) {
                            throw Exception("Too many ']' in text")
                        }
                        val start = startStack.removeLast()
                        val end = position
                        val span = TextSpan.fromBounds(start, end)
                        spanBuilder.add(span)
                    }
                    else -> {
                        position += 1
                        textBuilder.append(c)
                    }
                }
            }

            if (startStack.isNotEmpty()) {
                throw IllegalArgumentException("Too many '[' in text")
            }
            return AnnotatedText(textBuilder.toString(), spanBuilder)
        }

        fun unindentLines(text: String): List<String> {
            val lines = text.split("\r\n", "\n")
            val minIndentation = lines.filter { it.isNotBlank() }.minOf { it.length - it.trimStart().length }
            val dedentedLines = lines.map {
                if (it.isNotBlank()) {
                    it.substring(minIndentation)
                } else {
                    it
                }
            }
            val skipBegin = dedentedLines.takeWhile { it.isBlank() }.count()
            val skipEnd = dedentedLines.takeLastWhile { it.isBlank() }.count()

            return dedentedLines.subList(skipBegin, dedentedLines.size - skipEnd)
        }

        private fun unindent(text: String): String {
            return unindentLines(text).joinToString("\n")
        }
    }
}
